internal/storage/sql/common/flag.go

```diff
@@ -118,37 +130,92 @@ func (s *Store) ListFlags(ctx context.Context, opts ...storage.QueryOption) (sto
 		}
 	}()
 
+	// keep track of flags we've seen so we don't append duplicates because of the join
+	uniqueFlags := make(map[string][]*flipt.Variant)
+
 	for rows.Next() {
 		var (
-			flag      = &flipt.Flag{}
-			createdAt fliptsql.Timestamp
-			updatedAt fliptsql.Timestamp
+			f = &flipt.Flag{}
+			v = &variantRow{}
+
+			fCreatedAt fliptsql.Timestamp
+			fUpdatedAt fliptsql.Timestamp
 		)
 
 		if err := rows.Scan(
-			&flag.Key,
-			&flag.Name,
-			&flag.Description,
-			&flag.Enabled,
-			&createdAt,
-			&updatedAt); err != nil {
+			&f.Key,
+			&f.Name,
+			&f.Description,
+			&f.Enabled,
+			&fCreatedAt,
+			&fUpdatedAt,
+			&v.Id,
+			&v.Key,
+			&v.FlagKey,
+			&v.Name,
+			&v.Description,
+			&v.Attachment,
+			&v.CreatedAt,
+			&v.UpdatedAt); err != nil {
 			return results, err
 		}
 
-		flag.CreatedAt = createdAt.Timestamp
-		flag.UpdatedAt = updatedAt.Timestamp
+		f.CreatedAt = fCreatedAt.Timestamp
+		f.UpdatedAt = fUpdatedAt.Timestamp
 
-		if err := s.variants(ctx, flag); err != nil {
-			return results, err
+		// append flag to output results if we haven't seen it yet
+		if _, ok := uniqueFlags[f.Key]; !ok {
+			flags = append(flags, f)
```

Comment:
I believe you could actually do all of this without the map. Given we ensure the return rows are ordered by Flag. This might be the natural order of the join. However, we could ensure it in the `ORDER BY` clause. e.g. `ORDER BY f.created_at, f.key DESC`.

Once we have that invariant, we can simply do the following two steps:

1. Compare the newly scanned flag key to the last seen flag key in the result so far.
2. If the flag keys differ, then append the newly scanned flag to the results.
3. `append` the currently scanned variant to the last flag in the results seen so far.